library(numbat)
library(argparse)
library(dplyr)

parser <- ArgumentParser()
parser$add_argument('--donor', type = "character", required = TRUE, help = "patient")
args <- parser$parse_args()

donor <- args$donor

#all_counts <- readRDS('../anno/CRC/3p/all_epi_counts.rds')
all_counts <- readRDS('../../datasets_new_preprocessing/all_epi_cell_CB_counts.rds')
tumor_counts <- all_counts[,grepl(paste0(donor,'t'),colnames(all_counts))]
normal_counts <- all_counts[,grepl('n',colnames(all_counts))]

allele_df <- read.table(paste0('../../datasets_new_preprocessing/Numbat/',donor,'/',donor,'t_allele_counts.tsv.gz'),sep='\t',header=1)
allele_df$cell <- paste0(donor,'t:',gsub('-1$','',allele_df$cell))

#anno <- read.csv('../anno/CRC/3p/all_epi_anno.txt',sep='\t',header=0)
anno <- read.csv('../../datasets_new_preprocessing/all_epi_cell_anno.txt',sep=' ',header=1)
anno = anno %>% tibble::rownames_to_column('cell')
colnames(anno) = c('cell', 'group')

tumor_anno <- anno[grepl(paste0(donor,'t'),anno$cell),]
normal_anno <- anno[grepl('n',anno$cell),]

ref_internal <- aggregate_counts(normal_counts, normal_anno)

numbat_out <- run_numbat(
    tumor_counts, # gene x cell integer UMI count matrix
    ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix
    allele_df, # allele dataframe generated by pileup_and_phase script
    genome = "hg38",
    t = 1e-5,
    ncores = 1,
    plot = TRUE,
    out_dir = paste0('../../datasets_new_preprocessing/Numbat/', donor)
)

